### PR TITLE
Add EverParse based parser for ELF files

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Signed-off-by: Alan Jowett <alan.jowett@microsoft.com>

## Description

Invoke the EverParse build parser to validate ELF files prior to calling ebpf-verifier.

## Testing

CI/CD - bpf2c tests.

## Documentation

No.

Resolves: #834

### CODE REVIEW

File: libs/spec/spec.vcxproj [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+    <CustomBuildStep>
+      <Command>pushd $(OutDir)
+$(SolutionDir)packages\everparse\everparse.cmd $(ProjectDir)Elf.3d
+popd</Command>
+    </CustomBuildStep>
+    <CustomBuildStep>
+      <Outputs>$(OutDir)Elf.h;$(OutDir)ElfWrapper.h;$(OutDir)EverParse.h;$(OutDir)Elf.c;$(OutDir)\ElfWrapper.c;%(Outputs)</Outputs>
+    </CustomBuildStep>
+    <PreBuildEvent>
+      <Command>powershell -NonInteractive -ExecutionPolicy Unrestricted $(SolutionDir)scripts\Install-Tool.ps1 -URL 'https://github.com/project-everest/everparse/releases/download/v2022.05.10/everparse_v2022.05.10_Windows_NT_x86_64.zip' -PackageHash 'A6D14F80ED901DC41932C422D8FD07B8D695A463C9F688DCDC5E4764E1530BD3' -PackageName everparse -PackageRoot E:\ebpf-for-windows\packages</Command>
```
Comment (shankarseal) [inline libs/spec/spec.vcxproj:127]: why not do this in the CI job where the rest of the tools are downloaded.

Comment (Alan-Jowett) [reply libs/spec/spec.vcxproj:127]: Because it needs to be done on a developers box as well.

File: libs/spec/Elf.3d [hunk 2]
```diff
@@ -0,0 +1,460 @@
...
+             |                 |             |                  |
+             |                 |            \ /                \ /
+   ----------------------------------------------------------------------
+  |..|ph_offset|ph_num|..|sh_offset|sh_num|..|ph table|file data|sh table|
+   ----------------------------------------------------------------------
+
+  |<---------------ELF Header--------------->|
+
+  Where ph_table contains ph_num entries, sh_table contains sh_num entries,
+  and both describe sections in the rest of the file (i.e. in the file data part)
```
Comment (dthaler) [inline libs/spec/Elf.3d:32]:
```suggestion
  and both describe sections in the rest of the file (i.e., in the file data part)
```
nit

File: libs/spec/Elf.3d [hunk 3]
```diff
@@ -0,0 +1,460 @@
...
+
+  This specification validates the:
+  - ELF header
+  - Program Header table (that it contains ph_num entries,
+                          and points to valid parts of the rest of the file)
+  - Section Header table (that it contains sh_num entries,
+                          and points to valid parts of the rest of the file)
+
+  Though it is possible to do it in 3d, we do not validate individual sections/segments yet.
```
Comment (dthaler) [inline libs/spec/Elf.3d:41]: The PR description says it resolves 834.  Is it really sufficient to resolve it if we don't validate individual sections?

File: libs/spec/Elf.3d [hunk 4]
```diff
@@ -0,0 +1,460 @@
...
+
+  See the ELF type at the end, that's the entrypoint for the specification
```
Comment (dthaler) [inline libs/spec/Elf.3d:43]:
```suggestion
  See the ELF type at the end, that's the entrypoint for the specification.
```

File: libs/spec/Elf.3d [hunk 5]
```diff
@@ -0,0 +1,460 @@
...
+typedef UINT8 UCHAR;
+
+typedef struct _ZeroByte
+{
+    UINT8 zero { zero == 0 };
+} ZeroByte;
+
+// Address type is either UINT64 or UINT32 depending on the architecture
+// For now, the spec supports only 64 bit arch, later we could parameterize
+//   the Address type
```
Comment (dthaler) [inline libs/spec/Elf.3d:58]:
```suggestion
// the address type.
```

File: libs/spec/Elf.3d [hunk 6]
```diff
@@ -0,0 +1,460 @@
...
+
+// The same goes for the OFFSET type
```
Comment (dthaler) [inline libs/spec/Elf.3d:60]:
```suggestion
// The same goes for the OFFSET type.
```
nit: comments are inconsistent throughout this file

File: libs/spec/Elf.3d [hunk 7]
```diff
@@ -0,0 +1,460 @@
...
+}
+
+//
+// Last 7 bytes of the ELF header are 0s
+//
+
+#define E_IDENT_PADDING_SIZE    7
+
+
+/*++
```
Comment (dthaler) [inline libs/spec/Elf.3d:134]: why the ++ -- comment style? Does everparse do something special with such comments?

File: libs/spec/Elf.3d [hunk 8]
```diff
@@ -0,0 +1,460 @@
...
+  The e_ident array in the ELF header
+  --*/
+
+typedef struct _E_IDENT
+{
+  //The first four bytes are 0x7f followed by 'E', 'L', and 'F'
```
Comment (dthaler) [inline libs/spec/Elf.3d:140]: nit: comments are inconsistent in terms of whether a space appears after the // throughout this file

File: tests/bpf2c_tests/elf_bpf.cpp [hunk 9]
```diff
@@ -36,6 +36,7 @@ read_contents(const std::string& source, std::vector<std::function<std::string(c
     return return_value;
 }
 
+template <char seperator>
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply tests/bpf2c_tests/elf_bpf.cpp:39]: Need to revert this (was accidentally included).

File: libs/spec/Elf.3d (hunk 3 above)
Comment (Alan-Jowett) [reply libs/spec/Elf.3d:41]: #834 is about building the parse from the .3d file as part of the build.

A second effort will be required to extend the basic ELF parser to a BPF specific ELF parser (that checks the sections etc).

**Pull request closed without merging**